### DISCUSSION TITLE
HTTPS localhost / Nitro

### DISCUSSION THREAD

**Discussion by tereith** [nuxt/nuxt#16057] [Questions] [answered]
Hi! Nuxt noob here. I am trying to get https to work on localhost with the Nuxt3 Beta using Nitro. Anybody got some insight on this?

Specifically, where do I add the certificates?

Something ala this in Express:
const express = require('express');
const https = require('https');
const app = express();
const server = https.createServer({key: key, cert: cert }, app);

Regards,
Terje Eithun
Norway

**Comment by wobsoriano** [comment]
Try if this works

```ts
import { defineNuxtConfig } from 'nuxt3'
import { resolve } from 'path'
import { readFileSync } from 'fs'

export default defineNuxtConfig({
    server: {
        https: {
            key: readFileSync(resolve(__dirname, 'server.key')),
            cert: readFileSync(resolve(__dirname, 'server.crt'))
        }
    }
})
```

**Reply by tereith** [reply]
Hi, thanks for replying but it didn't work. I also got a Typescript error probably because according to the index.d.ts file the https property should be a boolean (I am confused by the 'example' VS the https: boolean):
```

  server: {
    /**
     * Whether to enable HTTPS.
     * @default false
     * 
     * @example
     * ```js
     * export default {
     *   server: {
     *     https: {
     *       key: fs.readFileSync(path.resolve(__dirname, 'server.key')),
     *       cert: fs.readFileSync(path.resolve(__dirname, 'server.crt'))
     *     }
     *   }
     * }
     * ```
    */
    https: boolean,

    /** @default 3000 */
    port: number,

    /** @default "localhost" */
    host: string,

    socket: any,

    /**
     * Enabling timing adds a middleware to measure the time elapsed during server-side rendering and adds it to the headers as 'Server-Timing'.
     * Apart from true/false, this can be an object for providing options. Currently, only `total` is supported (which directly tracks the whole time spent on server-side rendering.
    */
    timing: () => any,
  },
```

**Comment by alvarosabu** [comment]
Same here, I set it to `true` but still serves on HTTP

**Comment by pi0** [comment, marked as answer]
`server` options from `nuxt.config` are not supported with Bridge and Nuxt 3 anymore because `nuxi` development server starts earlier than loading config and production build is standalone not depending on nuxt config. 

In order to use the dev server with a self-signed certificate, you can use `nuxi dev --https` or provide keys with `--ssl-cert` and `ssl-key` for your own cert. (self-signed featured enabled in last build by #1699) [docs](https://v3.nuxtjs.org/getting-started/commands/#development-server)

**Reply by tereith** [reply]
Thanks, it works! Had to use both --https and key/cert. In my package.json:
 `"dev": "nuxi dev --https --ssl-cert localhost.pem --ssl-key localhost-key.pem"`
 
 But why does the page reload itself every few seconds?

**Reply by alvarosabu** [reply]
I had the same issue with the page being reloaded constantly, also after updating I got a weird warning 

```
[worker] Named export 'isWindows' not found. The requested module 'std-env' is a CommonJS module, which may not support all module.exports as named exports.
CommonJS modules can always be imported via the default export, for example using:

import pkg from 'std-env';
const { provider, isWindows } = pkg;


  import { provider, isWindows } from 'std-env';
  ^^^^^^^^^
  SyntaxError: Named export 'isWindows' not found. The requested module 'std-env' is a CommonJS module, which may not support all module.exports as named exports.
  CommonJS modules can always be imported via the default export, for example using:
  
  import pkg from 'std-env';
  const { provider, isWindows } = pkg;
  
  at ModuleJob._instantiate (node:internal/modules/esm/module_job:124:21)
  at async ModuleJob.run (node:internal/modules/esm/module_job:179:5)
  at async Loader.import (node:internal/modules/esm/loader:178:24)
  at async Object.loadESM (node:internal/process/esm_loader:68:5)
  at async handleMainPromise (node:internal/modules/run_main:63:12)
```

Nuxt stays on a infinite load state

**Reply by alvarosabu** [reply]
There is a work around here just in case somebody face this 

https://github.com/nuxt/framework/issues/1697

**Reply by davestewart** [reply]
> There is a work around here just in case somebody face this
> 
> #1697

Thanks! Regarding that workaround, the option that worked was installing `std-env`:

```bash
npm i std-env --save-dev
```

**Reply by shtse8** [reply]
> `server` options from `nuxt.config` are not supported with Bridge and Nuxt 3 anymore because `nuxi` development server starts earlier than loading config and production build is standalone not depending on nuxt config.
> 
> In order to use the dev server with a self-signed certificate, you can use `nuxi dev --https` or provide keys with `--ssl-cert` and `ssl-key` for your own cert. (self-signed featured enabled in last build by #1699) [docs](https://v3.nuxtjs.org/getting-started/commands/#development-server)

Can we set it in `.env` file?

**Reply by pi0** [reply]
Use a fresh lockfile please. It should be resolved.

**Reply by blowsie** [reply]
I think the docs need updating

**Comment by alvarosabu** [comment]
@tereith did you find a solution for the constant refreshing issues?

**Reply by tereith** [reply]
I am only testing Nuxt3 so I haven't tried very hard. Perhaps we have to wait for the Nuxt devs to fix this.
This kind of works for me (not in every browser):

try installing this: https://www.npmjs.com/package/@originjs/vite-plugin-commonjs

Then make a vite.config.ts containing:

```
import { viteCommonjs } from '@originjs/vite-plugin-commonjs'

export default {
    plugins: [
        viteCommonjs()
    ],
    server: {
        https: true,
        hmr: true
    }
}

```

**Reply by tereith** [reply]
Ahh.. the viteCommonjs part solved another problem for me so perhaps you don't need that. So in vite.config.ts:

```
export default {

    server: {
        https: true,
        hmr: true
    }
}

```

**Comment by multiplehats** [comment]
For anyone looking into this, this is how you solve it: https://github.com/nuxt/framework/issues/1796

**Comment by L422Y** [comment]
For local development, using the same cert (self signed, invalid even) for HTTPS and WSS and allowing an exception for the invalid certificate, Brave looks like it is applying the same exception for the WSS connection (same cert and domain, xyz.l0cal)

Obviously not ideal but it's the least annoying solve I've found yet.
